Event timing

goodhoko opened the pull request:
_note: I know our project is officially over but I just feel that a) this is getting really close to something useful b) I enjoy working on this so I may be submitting more PRs in upcoming days. I'll be glad for reviews but don't feel obliged in any way to respond to this. 🙌_

The first commit merges the two testing probes into one, taking inspiration from @PabloMansanet's "modes" in the [log probe](https://github.com/tonarino/acoustic_profiler/pull/38))

The second adds another mode to this unified probe that can be used to test out the last commit in a controlled way.

And that last commit implements a half-baked change that plays sounds for the incoming events based on their internal timestamp rather than as they come. It's only half-baked because I have a few questions I wanted to discuss first.

btw: this seems to work! Running the test probe in burst mode with arbitrary parameters results in "clacks" played evenly spaced.

goodhoko committed:
Merge tick probe with test probe

This will make adding more test modes easier.

goodhoko commented on crates/test_probe/src/main.rs:
```diff
@@ -37,41 +46,66 @@ fn main() -> Result<()> {
+        });
+
+        sleep(deadline.saturating_duration_since(Instant::now()));
+    }
```
As the code stands, all the events that are sent out at time `T` have timestamps `> T`. That's the opposite of what a typical real-world aggregating probe would produce. Such probe would be collecting events and then it would send them out with timestamps set in past.

I tried to mimic this by setting
```rust
let start_unix = SystemTime::now()
    .duration_since(UNIX_EPOCH)            👇
    .expect("Couldn't get unix timestamp") - burst_period;
```

What I'd expect then, is that once `burst_period` exceeds `AudioOutput`'s `play_delay` we'll start getting "too early plays" in every burst.

But that wasn't the case! Instead, only the first burst is "too early" and all subsequent bursts are played smoothly. I can't figure out how that's possible. 🤔

strohel replied:
:thinking:

goodhoko replied:
Aha! It was caused by a off-by-one bug in the burst for-loop. Fixed in https://github.com/tonarino/acoustic_profiler/pull/44. .)

goodhoko commented on crates/composer/src/jukebox.rs:
```diff
@@ -49,12 +49,17 @@ impl Jukebox {
 
-    pub(crate) fn play(&self, audio_output: &AudioOutput, sample: Sample) {
+    pub(crate) fn play(
+        &self,
+        audio_output: &AudioOutput,
+        sample: Sample,
+        timestamp: Option<Duration>,
+    ) {
         let buffer = self
             .samples
             .get(&sample)
             .expect("programmer error, all possible samples should be loaded");
 
-        audio_output.play(buffer.clone());
+        audio_output.play(buffer.clone(), timestamp);
```
I don't like how we need to pass the timestamp through the jukebox. Maybe we could gut the jukebox out even more and instead of `play` that takes the `audio_output` we could have
```rust
get_sound(sample: Sample) -> Buffer
```

and pass the returned buffer to the `audio_output` at the cal site?

strohel replied:
> ```rust
> get_sound(sample: Sample) -> Buffer
> ```

Yeah making Jukebox independent of playback would be much nicer! I think this can even return `-> impl Source<f32> + Send + 'static` so that we don't have to leak the `Buffer` internal type.

goodhoko replied:
I put that to side into an issue: https://github.com/tonarino/acoustic_profiler/issues/43

goodhoko commented on crates/composer_api/src/lib.rs:
```diff
@@ -15,8 +15,16 @@ pub const DEFAULT_SERVER_ADDRESS: &str = "localhost:8888";
 #[derive(Serialize, Deserialize, Debug)]
 pub enum Event {
     TestTick,
-    StdoutWrite { length: usize },
-    StderrWrite { length: usize },
+    TimedTestTick {
+        /// timestamp as Duration since the UNIX_EPOCH
+        ts: Duration,
+    },
+    StdoutWrite {
+        length: usize,
+    },
+    StderrWrite {
+        length: usize,
+    },
     LogStats(LogStats),
 }
```
I'm not sure if the timestamp should be encapsulated in (some) Events or if we should send it alongside the events eg. in a wrapping struct.

To me, the later seems nicer but it doesn't fit nicely with sending multiple timestamped events in a single message which we do in the [log probe](https://github.com/tonarino/acoustic_profiler/pull/38/files#diff-dc41a8d6981cf8fc8a37dfe62d23c2d48692271f844aefb407aaa7b9a1fa4486R30-R34).

strohel replied:
> To me, the later seems nicer

To me too. I even starter to write a comment above saying: At some point (not in this PR) we may want to add timestamp to all events, something like
```rust
pub struct Event {
    timestamp: Duration,
    kind: EventKind,
}

pub enum EventKind {
    TestTick,
    ...
}
```

> But it doesn't fit nicely with sending multiple timestamped events in a single message which we do in the [log probe](https://github.com/tonarino/acoustic_profiler/pull/38/files#diff-dc41a8d6981cf8fc8a37dfe62d23c2d48692271f844aefb407aaa7b9a1fa4486R30-R34).

Hmm, maybe we should have 2 distinct "higher-kind" event types, OneShot and Aggregate/Continuous or something like that. So `struct Event` could again become
```rust
enum Event {
    OneShot { timestmap: ..., kind: OneShotEventKind },
    Aggregate { ... }
}
```

goodhoko replied:
I really like how elegantly @skywhale solved this in #41 by *always* sending a Vec of optionally timed Events. `OneShot` is then just a regular message with a single event in it. .)

strohel commented on crates/composer/src/audio_output.rs:
```diff
@@ -66,9 +66,12 @@ impl AudioOutput {
+        source: S,
+        timestamp: Option<Duration>,
+    ) {
+        let play_at_timestamp = timestamp.unwrap_or_else(current_timestamp) + self.play_delay;
```
It feels slightly unRusty to have the poor man's `timestamp` default argument. When the default value is simple like this, I'd even make the argument mandatory and let caller handle that.

Another option would be to have `play_at()` that takes timestamp and `play()` that creates a default one and calls `play_at()`.

And most powerful would be to create some `struct PlayTask<S> { source: S, timestamp: Option<Duration> }` with factory methods, but that's really an overkill for our purposes.

strohel commented on crates/composer/src/main.rs:
```diff
@@ -59,15 +59,15 @@ fn handle_datagram(
+        Event::StdoutWrite { length: _ } => jukebox.play(audio_output, Sample::Click, None),
+        Event::TimedTestTick { ts } => jukebox.play(audio_output, Sample::Clack, Some(ts)),
     };
-    jukebox.play(audio_output, sample);
```
Under "duplicate only what you really need to duplicate" we would do something like:
```rust
let (sample, timestamp /* Option */) = match event {
    ...
};
jukebox.play(audio_output, sample, timestamp);

strohel commented on crates/composer_api/src/lib.rs:
```diff
@@ -15,8 +15,16 @@ pub const DEFAULT_SERVER_ADDRESS: &str = "localhost:8888";
-    StderrWrite { length: usize },
+    TimedTestTick {
+        /// timestamp as Duration since the UNIX_EPOCH
+        ts: Duration,
```
```suggestion
        timestamp: Duration,
```
Better to be explicit, especially in APIs.

strohel reviewed (commented):
Niiice! And well posed questions.

skywhale commented:
@goodhoko thank you so much for making progress on this front! I wanted to start testing timed events for other event types, too, so I created https://github.com/tonarino/acoustic_profiler/pull/41. It partially overlaps with your 3rd commit here, so please have a look. I've addressed some of @strohel's review comments, too.

goodhoko committed:
Expose current_timestamp util

Our particular way of expressing a timestamp is exposed as part of the
public Event type. There's no reason to not expose the accompanying
utility as well.

It will come in handy in the following commits.

goodhoko committed:
Fix Clippy warnings

goodhoko committed:
Add burst mode to test probe

Add a mode to the test probe for testing event ordering and
time-positioning of the composer in a controlled way.

Make sure the sequence of event timestamps progresses in a stable way
independent of the bursting rhythm which tends to hiccup. (Probably due
to varying execution times and imprecise thread::sleep)

goodhoko committed:
Fix off-by-one error

The `deadline` obtained from the iterator should hold the time of the
*next* iteration. If the first value of the iterator is equal to "now"
then the loop launches with two immediate iterations and only then
stabilizes on the desired frequency.

It goes:

iteration | deadline       | pause since last iteration
----------|----------------|---------------------------
first     | now            | n/a (just started)
second    | 1 period later | 0 (🐞 waited until "now")
third     | 1 period later | 1 period (back on track)
...

Start the period multiplier from 1 to fix this. Then we get:

iteration | deadline       | pause since last iteration
----------|----------------|---------------------------
first     | 1 period later | n/a (just started)
second    | 1 period later | 1 period (👈 fixed)
...

goodhoko commented:
I'll close this as the crux of this PR (timed events) was completed in #41 and #42.

I'll post the left-over bit in a clean PR.
